File: Problems/56/56_Merge_Intervals.py
# Cleaner solution
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort()
        res = [intervals[0]]
        for start, end in intervals[1:]:
            if start <= res[-1][1]:
                res[-1][1] = max(res[-1][1], end)
            else:
                res.append([start, end])
        return res

# class Solution:
#     def merge(self, intervals: List[List[int]]) -> List[List[int]]:
#         intervals.sort()
#         res = []
#         pre_start = intervals[0][0]
#         pre_end = intervals[0][1]
#         for start, end in intervals:
#             if start <= pre_end:
#                 pre_end = max(pre_end, end)
#             else:
#                 res.append([pre_start, pre_end])
#                 pre_start = start
#                 pre_end = end
#         res.append([pre_start, pre_end])
#         return res

File: Problems/56/56_Merge_Intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end());
        vector<vector<int>> res;
        res.push_back(intervals[0]);
        for (int i = 1; i < intervals.size(); ++i) {
            int start = intervals[i][0];
            int end = intervals[i][1];
            if (start <= res.back()[1]) {
                res.back()[1] = max(res.back()[1], end);
            } else {
                res.push_back({start, end});
            }
        }
        return res;
    }
};